FreeRTOS-Plus/Demo/AWS/Fleet_Provisioning_Windows_Simulator/Fleet_Provisioning_With_CSR_Demo/pkcs11_operations.c

```diff
@@ -0,0 +1,1038 @@
+    CK_KEY_TYPE xKeyType = CKK_EC;
+
+    CK_BBOOL xTrueObject = CK_TRUE;
+    CK_ATTRIBUTE pxPublicKeyTemplate[] =
+    {
+        { CKA_KEY_TYPE,  NULL /* &keyType */,         sizeof( xKeyType )         },
+        { CKA_VERIFY,    NULL /* &trueObject */,      sizeof( xTrueObject )      },
+        { CKA_EC_PARAMS, NULL /* ecParams */,         sizeof( pxEcParams )       },
+        { CKA_LABEL,     ( void * ) pcPublicKeyLabel, strlen( pcPublicKeyLabel ) }
+    };
+
+    /* Aggregate initializers must not use the address of an automatic variable. */
+    pxPublicKeyTemplate[ 0 ].pValue = &xKeyType;
+    pxPublicKeyTemplate[ 1 ].pValue = &xTrueObject;
+    pxPublicKeyTemplate[ 2 ].pValue = &pxEcParams;
+
+    CK_ATTRIBUTE privateKeyTemplate[] =
+    {
+        { CKA_KEY_TYPE, NULL /* &keyType */,          sizeof( xKeyType )          },
+        { CKA_TOKEN,    NULL /* &trueObject */,       sizeof( xTrueObject )       },
+        { CKA_PRIVATE,  NULL /* &trueObject */,       sizeof( xTrueObject )       },
+        { CKA_SIGN,     NULL /* &trueObject */,       sizeof( xTrueObject )       },
+        { CKA_LABEL,    ( void * ) pcPrivateKeyLabel, strlen( pcPrivateKeyLabel ) }
+    };
+
+    /* Aggregate initializers must not use the address of an automatic variable. */
+    privateKeyTemplate[ 0 ].pValue = &xKeyType;
+    privateKeyTemplate[ 1 ].pValue = &xTrueObject;
+    privateKeyTemplate[ 2 ].pValue = &xTrueObject;
+    privateKeyTemplate[ 3 ].pValue = &xTrueObject;
+
+    xResult = C_GetFunctionList( &xFunctionList );
+
+    if( xResult != CKR_OK )
+    {
+        LogError( ( "Could not get a PKCS #11 function pointer." ) );
+    }
+    else
+    {
+        xResult = xFunctionList->C_GenerateKeyPair( xSession,
+                                                    &xMechanism,
+                                                    pxPublicKeyTemplate,
+                                                    sizeof( pxPublicKeyTemplate ) / sizeof( CK_ATTRIBUTE ),
+                                                    privateKeyTemplate, sizeof( privateKeyTemplate ) / sizeof( CK_ATTRIBUTE ),
+                                                    xPublicKeyHandlePtr,
+                                                    xPrivateKeyHandlePtr );
+    }
+
+    return xResult;
+}
+
+/*-----------------------------------------------------------*/
+
+bool xGenerateKeyAndCsr( CK_SESSION_HANDLE xP11Session,
+                         const char * pcPrivKeyLabel,
+                         const char * pcPubKeyLabel,
+                         char * pcCsrBuffer,
+                         size_t xCsrBufferLength,
+                         size_t * pxOutCsrLength )
+{
+    CK_OBJECT_HANDLE xPrivKeyHandle;
+    CK_OBJECT_HANDLE xPubKeyHandle;
+    CK_RV xPkcs11Ret = CKR_OK;
+    mbedtls_pk_context xPrivKey;
+    mbedtls_pk_info_t xPrivKeyInfo;
+    mbedtls_ecdsa_context xEcdsaContext;
+    mbedtls_x509write_csr xReq;
+    int32_t usMbedtlsRet = -1;
```

Comment:
This variable isn't named correctly.

Reply:
Addressed in this commit: https://github.com/FreeRTOS/FreeRTOS/pull/724/commits/1cd54dafc3db6ce1736300bd087df9488bbd3437